de.peeeq.wurstscript/src/main/java/de/peeeq/wurstscript/intermediatelang/optimizer/BranchMerger.java

```diff
@@ -0,0 +1,109 @@
+
+import com.google.common.collect.Lists;
+import de.peeeq.wurstscript.intermediatelang.optimizer.ControlFlowGraph.Node;
+import de.peeeq.wurstscript.jassIm.*;
+import de.peeeq.wurstscript.translation.imtranslation.ImTranslator;
+import org.eclipse.jdt.annotation.Nullable;
+
+import java.util.List;
+
+/**
+ * merges identical nodes in branches if possible without side effects
+ * <p>
+ * the input must be a flattened program
+ */
+public class BranchMerger {
+    public int totalLocalsMerged = 0;
+    private final ImProg prog;
+    private final ImTranslator trans;
+
+    public BranchMerger(ImTranslator trans) {
+        this.prog = trans.getImProg();
+        this.trans = trans;
+    }
+
+    public void optimize() {
+        for (ImFunction func : prog.getFunctions()) {
+            optimizeFunc(func);
+        }
+    }
+
+    private void optimizeFunc(ImFunction func) {
+        mergeBranches(func);
+    }
+
+
+    private void mergeBranches(ImFunction func) {
+        ControlFlowGraph cfg = new ControlFlowGraph(func.getBody());
+
+        // init in and out with empty sets
+        for (Node n : cfg.getNodes()) {
+            @Nullable ImStmt stmt = n.getStmt();
+            if (stmt != null && stmt.getParent() != null && stmt.getParent() instanceof ImIf) {
+                ImIf imIf = (ImIf) stmt.getParent();
+                if (n.getPredecessors().size() <= 1 && n.getSuccessors().size() == 2) {
+                    Node leftStmt = n.getSuccessors().get(0);
+                    Node rightStmt = n.getSuccessors().get(1);
+                    if (leftStmt.toString().equals(rightStmt.toString())) {
+                        if (n.getPredecessors().size() == 1) {
+                            // Possible match. At last check if condition causes sideeffects.
+                            if (hasNoSideEffects(n, leftStmt)) {
+
+                                ImStmt mergedStmt = leftStmt.getStmt().copy();
+                                ImIf oldIf = imIf.copy();
+                                oldIf.getThenBlock().get(0).replaceBy(JassIm.ImNull());
+                                oldIf.getElseBlock().get(0).replaceBy(JassIm.ImNull());
+                                imIf.replaceBy(JassIm.ImStatementExpr(JassIm.ImStmts(mergedStmt, oldIf), JassIm.ImNull()));
+                            }
+                        }
+                    }
+                }
+                // if the first statement of each branch of the if is the same, it can be moved before the branch
+            }
+        }
+    }
+
+
+    private List<ImVar> calculateVarUses(Node node) {
+        List<ImVar> result = Lists.newArrayList();
+        ImStmt stmt = node.getStmt();
+        if (stmt != null) {
+            stmt.accept(new ImStmt.DefaultVisitor() {
+                @Override
+                public void visit(ImFunctionCall call) {
+                    super.visit(call);
+                    visit(call.getFunc());
```

Comment:
To avoid endless loops you can use a similar approach like the StackTraceInjector:
```
Multimap<ImFunction, ImFunction> callRelationTr = Utils.transientClosure(callRelation);
```
This will give you a Multimap, where `callRelationTr.get(f))` will give you all functions called by `f` and also those called by functions called by `f` and so on (transitive closure).

With that it should be easier to calculate the set of variables that might be changed by each function.